### PR TITLE
#30 & #31: Add scaleFactorMax and scaleFactorMin props. Add onZoom and onPan…

### PR DESCRIPTION
**Pull request by brentoneill**
… callback props.

I did not run the builds as I didn't want to clutter this PR, but I can run them once the changes are approved.

Let me know what you think! It's working pretty well for our application now :)

### CODE REVIEW

**Comment by brentoneill** [comment]
Considering adding a `onAuto` callback to this PR that fires when using the auto tool...

Thoughts?

**Comment by chrvadala** [comment]
I guess that this feature can be easily produced with a callback on `onChangeValue`

```
onChangeValue(()=>{
  if(tool !== TOOL_AUTO) return

  //code
})
```

**Comment by brentoneill** [comment]
Ahhh that makes sense. I'll try that out.

**Comment by brentoneill** [comment]
Anything else we need to do to get this PR merged? Some tests maybe?

**Comment by chrvadala** [comment]
I don't see any change from my last comment. Have you pushed your code?

**Comment by brentoneill** [comment]
Confused, thought your last comment was just a suggestion on how to get that same result - which it worked. I ended up using the `onChangeValue` callback on our implementation.

**Comment by chrvadala** [comment]
Sorry, not my last comment, but the previous one. I added few comments on the code...
https://github.com/chrvadala/react-svg-pan-zoom/pull/34#pullrequestreview-31292110

**Comment by brentoneill** [comment]
All looks good to me! @chrvadala

**Comment by chrvadala** [comment]
I've just found why you didn't see my review... I wrote it one week ago, but Github needs an explicit confirm and so it was in pending state. Now I guess that you can see it. 
If there are some doubts or difficulties send me a comment!
Thanks for your support and sorry for this mistake

**Comment by brentoneill** [comment]
I see the changes requested now, thanks! I will hop on those.

File: src/features/interactions.js [hunk 1]
```diff
@@ -29,11 +36,21 @@ export function onMouseDown(event, ViewerDOM, tool, value, props, coords = null)
   switch (tool) {
     case TOOL_ZOOM_OUT:
       let SVGPoint = getSVGPoint(value, x, y);
-      nextValue = zoom(value, SVGPoint.x, SVGPoint.y, 1 / props.scaleFactor);
+      if (props.scaleFactorMin && lessThanScaleFactorMin(props, nextValue)) {
+        console.error('Cannot zoom out past minimum scale factor of: ', props.scaleFactorMin);
```
Comment (brentoneill) [inline src/features/interactions.js]: These errors would only show if the scaleFacorMin or scaleFactorMax were provided, see the check for `props.scaleFactorMin`. Can still remove if desired.

File: src/viewer.js [hunk 2]
```diff
@@ -207,19 +207,25 @@ export default class ReactSVGPanZoom extends React.Component {
...
+            console.log(operation);
+            if(operation === ACTION_ZOOM) props.onZoom && props.onZoom(eventFactory(event, nextValue, this.ViewerDOM));
           }}
           onMouseMove={ event => {
             let {left, top} = this.ViewerDOM.getBoundingClientRect();
             let x = event.clientX - Math.round(left);
             let y = event.clientY - Math.round(top);
 
-            let nextValue = onMouseMove(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props, {x, y});
+            let {operation, nextValue} = onMouseDown(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
```
Comment (chrvadala) [inline src/viewer.js]: Is it a copy/paste mistake ?

File: src/viewer.js [hunk 3]
```diff
@@ -207,19 +207,25 @@ export default class ReactSVGPanZoom extends React.Component {
           style={cursor ? {cursor, display: "block"} : {display: 'block'}}
 
           onMouseDown={ event => {
-            let nextValue = onMouseDown(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
+            let {operation, nextValue} = onMouseDown(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
             if (this.getValue() !== nextValue) this.setValue(nextValue);
             this.handleViewerEvent(event);
+
+            console.log(operation);
```
Comment (chrvadala) [inline src/viewer.js]: remove it

File: src/viewer.js [hunk 4]
```diff
@@ -207,19 +207,25 @@ export default class ReactSVGPanZoom extends React.Component {
...
             let x = event.clientX - Math.round(left);
             let y = event.clientY - Math.round(top);
 
-            let nextValue = onMouseMove(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props, {x, y});
+            let {operation, nextValue} = onMouseDown(event, this.ViewerDOM, this.getTool(), this.getValue(), this.props);
             if (this.getValue() !== nextValue) this.setValue(nextValue);
             this.setState({viewerX: x, viewerY: y});
             this.handleViewerEvent(event);
+
+            console.log(operation);
```
Comment (chrvadala) [inline src/viewer.js]: please remove it

File: src/viewer.js (hunk 2 above)
Comment (4chen500) [inline src/viewer.js]: @chrvadala Is this typo the only thing preventing merge of this PR? I'd really love to use it. Thanks for your hard work!

**Comment by chrvadala** [comment]
Hi @brentoneill,
I've just released and merged all the recent features that has been implemented in ReactSVGPanZoom.
Could you take care of this PR and adapt it to this recent changes?

I'd like to merge it in the next version.
Thanks a lot!

**Comment by 4chen500** [comment]
@brentoneill @chrvadala 

Are you going to wrap up this PR? If you don't have the time, I'd be happy to take it over, as I'd really like to use this functionality. It really appears to be 98% complete...

**Comment by chrvadala** [comment]
Hi @4chen500,
I guess that @brentoneill closed this PR because he hasn't enough time to work on it. If you wanted to work on these features, I suggest to split #30 and #31 into two different PR.
 #31  should be extractable from this PR without too much work. #30 requires a lot of features not yet programmed.

**Comment by brentoneill** [comment]
I'd be happy to finish #31, but I closed this because I was not going to be able to fit in #30 any time soon. 

@4chen500 - please feel free to tie this up.

**Comment by hatpick** [comment]
So, is there any callback for zoom (in/out), pan, and fit to the viewer?

**Comment by chrvadala** [comment]
I've answered here https://github.com/chrvadala/react-svg-pan-zoom/issues/60#issuecomment-345296229

File: src/viewer.js (hunk 2 above)
Comment (auroranil) [reply src/viewer.js]: Bump

Comment (auroranil) [reply src/viewer.js]: Same here, would like to use it as well.

**Pull request closed without merging**